merkletree/merkletree.go

```diff
@@ -0,0 +1,288 @@
+	// subTrees to create a subTree with a height that is 1 greater.
+	for t.head != nil && current.height == t.head.height {
+		// When creating a proof for a specific index, you need to collect one
+		// hash at each height of the tree, and that hash will be found in the
+		// same subTree as the initial leaf. Before we hit that index, this
+		// logic will be ignored because len(proveSet) will be 0. After we hit
+		// that index, len(proveSet) will be one. From that point forward,
+		// every time there are two subTrees (the current one and the previous
+		// one) that have a height equal to len(proveSet) we will need to grab
+		// one of the roots and add it to the prove set.
+		if current.height == len(t.proveSet) {
+			// Either the root of the current subTree or the root of the
+			// previous subTree needs to be added to the proof set. We want to
+			// grab the root of the subTree that does not contain
+			// 't.proveIndex'. We do this by finding the starting index of the
+			// current subTree and comparing it to 't.proveInex'.
+			//
+			// The start of the first subTree can be determined by rounding
+			// the currentIndex down to the nearest (2^height). This represents
+			// the combined size of the two trees, as a tree of height 1 was
+			// built from only 1 leaf.
+			combinedSize := int(1 << uint(current.height))
+			previousStart := (t.currentIndex / combinedSize) * combinedSize
+			currentStart := previousStart + (combinedSize / 2)
+			if t.proveIndex < currentStart {
+				t.proveSet = append(t.proveSet, current.sum)
+			} else {
+				t.proveSet = append(t.proveSet, t.head.sum)
+			}
+		}
+
+		// Join the two subTrees into one subTree with a greater height. Then
+		// compare the new subTree to the next subTree.
+		current.sum = join(t.hash, t.head.sum, current.sum)
+		current.height++
+		t.head = t.head.next
+	}
+
+	// Add the subTree to the top of the stack.
+	current.next = t.head
+	t.head = current
+	t.currentIndex++
+}
+
+// Root returns the Merkle root of the data that has been pushed into the Tree.
+// Asking for the root when no data has been added will return nil. The tree is
+// left unaltered.
+func (t *Tree) Root() (root []byte) {
+	// If the Tree is empty, return nil.
+	if t.head == nil {
+		return nil
+	}
+
+	// The root is formed by hashing together subTrees in order from least in
+	// height to greatest in height. To preserve ordering, the larger subTree
+	// needs to be first in the combination.
+	current := t.head
+	root = current.sum
+	for current.next != nil {
+		root = join(t.hash, current.next.sum, root)
+		current = current.next
+	}
+	return root
+}
+
+// Prove returns a proof that the data at index 'proveIndex' is an element in
+// the current Tree. The proof will be invalid if any more elements are added
+// to the tree after calling Prove. The tree is left unaltered.
+func (t *Tree) Prove() (h hash.Hash, merkleRoot []byte, proveSet [][]byte, proveIndex int, numLeaves int) {
```

Comment:
Yeah. It's not strictly necessary, but I figured it should give you all of the inputs to verify.